workflows/mag.nf

```diff
@@ -423,118 +433,122 @@ workflow MAG {
     ================================================================================
     */
 
-    // Co-assembly: prepare grouping for MEGAHIT and for pooling for SPAdes
-    if (params.coassemble_group) {
-        // short reads
-        // group and set group as new id
-        ch_short_reads_grouped = ch_short_reads_assembly
-            .map { meta, reads -> [ meta.group, meta, reads ] }
-            .groupTuple(by: 0)
-            .map { group, metas, reads ->
-                def assemble_as_single = params.single_end || ( params.bbnorm && params.coassemble_group )
-                def meta         = [:]
-                meta.id          = "group-$group"
-                meta.group       = group
-                meta.single_end  = assemble_as_single
-                if ( assemble_as_single ) [ meta, reads.collect { it }, [] ]
-                else [ meta, reads.collect { it[0] }, reads.collect { it[1] } ]
-            }
-        // long reads
-        // group and set group as new id
-        ch_long_reads_grouped = ch_long_reads
-            .map { meta, reads -> [ meta.group, meta, reads ] }
-            .groupTuple(by: 0)
-            .map { group, metas, reads ->
-                def meta = [:]
-                meta.id          = "group-$group"
-                meta.group       = group
-                [ meta, reads.collect { it } ]
-            }
-    } else {
-        ch_short_reads_grouped = ch_short_reads_assembly
-            .filter { it[0].single_end }
-            .map { meta, reads -> [ meta, [ reads ], [] ] }
-            .mix (
-                ch_short_reads_assembly
-                    .filter { ! it[0].single_end }
-                    .map { meta, reads -> [ meta, [ reads[0] ], [ reads[1] ] ] }
-            )
-        ch_long_reads_grouped = ch_long_reads
-    }
+    if ( !params.assembly_input ) {
```

Comment:
Here the steps are skipped with if statements. That is clean, because the terminal output will not show those skipped processes, but it leads to such convoluted if else stuff. While I typically prefer the if statements as well, here it might be better to work with empty channels to skip steps. I didnt think the idea through, but if it would make the code much easier to read, I would rather prefer to use here empty-channel-skipping than if-skipping.

Reply:
Alright. I wont have time to review that current mess though, so go with it, you have James' ok.